plugins/README.adoc

```diff
@@ -7,3 +7,247 @@ NOTE: Work in Progress...
+
+Examples::
+[source,python]
+----
+def greeting(name: str) -> str:
+    return 'Hello ' + name
+----
+
+====
+
+== The use of unittest is discouraged, use pytest instead.
+[%collapsible]
+====
+Explanations:: Use pytest for writing unit tests for plugins
+
+Rationale:: Pytest is the testing framework used by Ansible Engineering and will provide the best experience for plugin developers
+
+Examples::
+[source,python]
+----
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import pytest
+
+from ansible.modules.copy import AnsibleModuleError, split_pre_existing_dir
+from ansible.module_utils.basic import AnsibleModule
+
+ONE_DIR_DATA = (('dir1',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ('dir1/',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ) 
+
+@pytest.mark.parametrize('directory, expected', ((d[0], d[2]) for d in ONE_DIR_DATA))
+def test_split_pre_existing_dir_one_level_exists(directory, expected, mocker):
+    mocker.patch('os.path.exists', side_effect=[True, False, False])
+    split_pre_existing_dir(directory) == expected
+
+----
+
+====
+
+== A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.
+[%collapsible]
+====
+Explanations::
+A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.  ( >30 keys or depth>2) (action plugins and control node execution modules)
+
+Rationale::
+Hand-crafting large and complex argument specifications can be error-prone and difficult to maintain.  Programmatically generating the spec from DOCUMENTATION (already a structured data object) ensures consistency and accuracy. 
+
+====
+
+== A plugin’s entry file should not contain more than 500 lines
+[%collapsible]
+====
+Explanations::
+A plugin’s entry file should not contain more than 500 lines
+
+Rationale::
+Long and complex code files can be difficult to maintain.
+Wherever possible, reusable functions and classes, such as those for data validation or manipulation, should be moved to module_utility or plugin_utility files and imported into plugins.
+This keeps the plugin code easier to read and maintain.
+500 lines is a general guideline that should serve most cases.
+====
+
+== Action plugins should be used when the code runs on the control node.
+[%collapsible]
+====
+Explanations::
+Use an action plugin for code that needs to run on the controller, where the host would be `localhost`.
+Rationale::
+Action plugins let you integrate local processing and local data with module functionality.
+Action plugins act in conjunction with modules to execute the actions required by playbook tasks.
+Using an action plugin will cause Ansible to skip creating a temporary tarball of execution code and runs the code directly on the controller host.
```

Comment:
Should this be enforced for platforms where the ability to skip this is already present (aka `import_modules`)? What other benefits does implementing as an action plugin have in such cases?

Reply:
Are there changes we should make to this section?  I'm happy to update the verbiage or add/remove things as needed - I'm less familiar with the networking use-case discussed so I'm not sure if there's changes needed.

Reply:
@NilashishC @cidrblock Thinking of this PR in the context of general guidelines (not, for example, Network team coding practices or platform specific edge cases), what changes should we make here?  Or is the section OK as-is?